### PR TITLE
gitserver: add GitserverLocalClone store

### PR DESCRIPTION
**Pull request by asdine**
This adds a simple store that will be used to enqueue clone jobs between gitserver shards.
The enqueue method will be used by several components:
- The OOB rendezvous migration, to migrate batches of repos
- The repository pinning feature, to trigger a migration whenever a repo is pinned

## Test plan

Wrote a unit test

Related to #32827

### CODE REVIEW

**Review by sashaostrikov — APPROVED** [review]
LGTM!

I've added a couple of questions and 1 little commit suggestion

File: internal/database/gitserver_localclone_jobs.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+	"github.com/sourcegraph/sourcegraph/internal/database/basestore"
+	"github.com/sourcegraph/sourcegraph/internal/database/dbutil"
+)
+
+// GitserverLocalCloneStore is used to migrate repos from one gitserver to another asynchronously.
+type GitserverLocalCloneStore interface {
+	basestore.ShareableStore
+	With(other basestore.ShareableStore) GitserverLocalCloneStore
+	Enqueue(ctx context.Context, repoID int, sourceHostname, destHostname string, deleteSource bool) (int, error)
```
[the remark this answers is not in the document]
Comment (asdine) [reply internal/database/gitserver_localclone_jobs.go:17]: It is actually very common but mostly used for optional parameters. For example:

```go
func (someType) List(ctx context.Context, opts *ListOptions) { }

// Call with option A
foo.List(ctx, &ListOptions{A: 1})

// Call with option B
foo.List(ctx, &ListOptions{B: 1})

// Call with option A and B
foo.List(ctx, &ListOptions{A: 1, B: 1})

// No option uses the default behaviour
foo.List(ctx, nil)
```

In our case, the first 4 parameters are mandatory, using a struct would require more validation steps, comments to indicate what is required etc. 
The last one is not though, so if we ever add more optional params, we will probably use the pattern you mentioned.

File: internal/database/gitserver_localclone_jobs.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+var _ GitserverLocalCloneStore = (*gitserverLocalCloneStore)(nil)
```
[the remark this answers is not in the document]
Comment (asdine) [reply internal/database/gitserver_localclone_jobs.go:20]: This is usually useful because it prevents the code from compiling if the gitserverLocalCloneStore doesn't implement the GitserverLocalCloneStore interface.
But in our case, I realize that it is not really useful because our constructor (NewGitserverLocalCloneStore) returns an interface, not a concrete type. I think I'll just delete this 👍🏼

File: internal/database/gitserver_localclone_jobs_test.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+func TestGitserverLocalCloneEnqueue(t *testing.T) {
+	if testing.Short() {
+		t.Skip()
+	}
+
+	db := dbtest.NewDB(t)
+	ctx := context.Background()
+
+	jobid, err := GitserverLocalClone(db).Enqueue(ctx, 1, "gitserver1", "gitserver2", true)
```
Comment (vrto) [inline internal/database/gitserver_localclone_jobs_test.go:18]: I'm just trying to understand this a little better, but it looks like we assume the data is in the DB thanks to the way we created `dbtest.NewDb(t)` above? So as a reader, I gotta dig in and find the dataset? :)

**Review by vrto — APPROVED** [review]
So far so good :)

File: internal/database/gitserver_localclone_jobs_test.go (hunk 3 above)
Comment (asdine) [reply internal/database/gitserver_localclone_jobs_test.go:18]: Sorry I'm not sure to understand the question, could you elaborate?
Basically we are using "stores" as interfaces between the caller and the database.
It is assumed that the caller doesn't have to know the dataset exactly to interact with the database.

Comment (vrto) [reply internal/database/gitserver_localclone_jobs_test.go:18]: Sure, let me rephrase. In this test code, something has to populate the test database and I was wondering where do data fixtures (sample data) come from.

Comment (vrto) [reply internal/database/gitserver_localclone_jobs_test.go:18]: Gotcha, thanks for the info! 👍
